### ISSUE TITLE
Any plans to implement/allow HEAD for API requests that allow GET?

### ISSUE THREAD

**Issue by amn** [hashicorp/vault#18224]
**Is your feature request related to a problem? Please describe.**
I have verified empirically that a HTTP `GET` request to `/v1/{secret-engine-mount}/data/{secret-name}` works, while the corresponding `HEAD` does not where Vault server returns 405. Admittedly, this may be due to [default] capabilities configured on the server -- I am running `vault server -dev`.I think in general a HTTP server could always trivially and as a rule support `HEAD` method for requests for which `GET` is allowed. That would just be a client not needing a response body, normally. Saves on the overhead for both the server and the client. Besides it's a trivial thing to implement, arguably -- at the simplest, just omit the body if the method is `HEAD`. Use cases include client wanting to determine if a secret (with given path) exists (`2xx` HTTP status code) or doesn't (HTTP status code `404`).

**Describe the solution you'd like**
Implement (or allow) `HEAD` requests for those resources for which `GET` is allowed.

**Describe alternatives you've considered**
Hard to say -- I want to determine if a secret that corresponds to a given secret path, exists -- I could issue a `GET` but ignore the response, but if the secret is large, that's overhead for transmission of the entire secret, not to mention a security factor (if I didn't actually need to read the secret value), also in cases where a principal may be allowed to check whether a secret exists but not actually _access_ (read) it.

**Explain any additional use-cases**
No, the above should pretty much cover it.

**Additional context**
No additional context, to the best of my understanding.

**Comment by maxb** [comment]
(Note: I'm not a HashiCorp employee, just an interested 3rd party Vault user)

What practical benefit do you seek to derive from the proposed feature, though?

You've made mention of "overhead", but let's be realistic here - JSON marshalling and network transmission of small amounts of data (and the vast majority of Vault data is small) is not usually a major overhead. You've proposed that it might be trivial to implement - but it's only trivial to implement if it's a rather basic implementation that does all the request processing, and then throws away the result, resulting in fairly limited benefit.

It's also necessary to consider that there are Vault endpoints - e.g. `ad/creds/:role_name` which conditionally perform write operations to backend systems during a `GET` request - if changing a password in an external system is needed to initially obtain it or comply with configured TTLs. Since Vault is a plugin architecture, it's difficult to enumerate which `GET` requests actually might cause internal side-effects - there are probably more in plugins I'm not familiar with.

You've implied near the end of your post, that you'd like to see it be possible to write policy that allows `HEAD` but not `GET` - that's another way this stops being trivial.


In conclusion, then, I think this issue is unlikely to go anywhere without more details of concrete use-cases for the proposed feature.

**Comment by amn** [comment]
Appreciate the elaborate response and explanations, thank you. This is not an issue I consider critical by any stretch -- certainly in practice I can get by issuing `GET` requests and ignore/discard the response body, in absence of support for the `HEAD` method.

About my use case: I am implementing a very simple Linux file system over [FUSE](https://en.wikipedia.org/wiki/Filesystem_in_Userspace) that maps Vault secrets to folders and secret key-value pairs to files. The motivation behind the project is the basically depending on the "everything is a file"-paradigm as we have services that are written to require access to secrets as files. For obvious security reasons we want to avoid storing secrets as files on persistent storage (i.e. your "regular" file system) specifically, opting for a FUSE-mounted "virtual" file system instead with Vault being the [secure] backing storage there.

As the file system responds to requests with arbitrary file/folder paths assumed to map to a secret or a key-value pair under one, practically, with `HEAD` I'd be able to just determine whether a corresponding secret even exists with a given path -- for some path requests I don't need the _data_ or much of _metadata_, only to determine whether there _is or isn't_ a corresponding secret in the Vault.

Like I said, this isn't a critical issue, I am merely taking the opportunity to elaborate.

I did not understand what you tried to describe with `GET` requests that may have internal side effects, but I should mention that HTTP requires `GET` to be [_idempotent_](https://httpwg.org/specs/rfc9110.html#idempotent.methods). This property should greatly help with any invariants during request processing that should in turn help with implementing processing of corresponding `HEAD`, at "no extra charge", ideally. But this of course depends on how Vault implements request processing in the first place, admittedly.

Regarding policy, and this is a digression -- I had not found out whether Vault *prevents* me from doing `HEAD` requests that it otherwise supports, because the method isn't "whitelisted", or whether the latter method isn't supported in the first place. That's what prompted the speculation, but I admit that even if implemented, it would make sense to control this through policy -- `HEAD` doesn't have to require actual access to the resource the way a `GET` request implies, so one could have finer-grained security controls with that.

**Comment by maxb** [comment]
Thanks for all that additional information.

I think a critical decision for you will be to decide whether you only need this to work for the Vault KV secrets engine, or if you are trying to make something that will cater for other parts of the Vault API as well.

The Vault KV secrets engine is at least somewhat similar to a filesystem. Not entirely - here are a few notable differences:
* You can have both a "file" and a "directory" of the same name, differentiated only by the presence or absence of a trailing slash
* "Directories" have no real existence of their own. There's no such thing as an empty directory, as they immediately vanish when the last contained file is removed. Nor can you create an empty directory, they spontaneously appear when a file with a suitable path is being created.
* The "list" operation exclusively lists directories. If you list "foobar", Vault will transparently add on an extra trailing slash, and give you the results as if you listed "foobar/" - even if "foobar/" doesn't exist.

Other parts of the Vault API are even less like a filesystem. There are write-only endpoints for example, and the discovery APIs for finding out what APIs exist tend to require human intelligence to interpret - not something that can be easily processed into a directory listing.

----

**On idempotency**

I'm well aware of what the HTTP spec says, but Vault arguably breaks this for some endpoints.

If I make the request `GET /v1/ad/creds/foobar`, Vault may reach out to my Active Directory server and change the password of the `foobar` account. It would be *surprising* if a `HEAD` request triggered that, which it would, in the "request processing as normal but discard the result" implementation of `HEAD` requests.

**Comment by raskchanky** [comment]
Hi @amn. There are no plans to implement anything like this currently. I will note, just as a data point, that as of this writing [sys/health](https://developer.hashicorp.com/vault/api-docs/system/health) does support HEAD, so it's not something we're opposed to philosophically. I do think the points that @maxb brings up are good ones. Were we to add support for HEAD to more of Vault, it would likely happen on a case by case basis, rather than as a blanket move to every endpoint that supports GET.

**Comment by hsimon-hashicorp** [comment]
Since an answer has been provided, I'll go ahead and close this issue for now, but please feel free to re-open it or continue the conversation as need be. Thanks! :)
